# Content list

1. Import libraries
2. Import data
3. Grouping data and aggregating variables
4. Export Data

# 01. Import libraries

In [51]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import data

In [52]:
# Create path
path = r'C:\Users\adrig\OneDrive\Documentos\CARRIERFOUNDRY\DATA IMMERSION\ACHIEVEMENT 4\Ex_4.2_Python\03-2025 Instacart Basket Analysis'

In [53]:
# Import data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new_var.pkl'))

# 03. Grouping data and aggregating variables

## 03.2 Average orders by department

In [54]:
# First split data into groups, then apply the aggregation function
ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## 03.3 Analysis from results

The results are very similar to the ones from the subset, just change in the decimal numbers. The department with highest average orders is the department_id: 21, followed by 10 and 18. 

## 03.4 Loyalty flag for customers

In [55]:
# Split data based on user_id, apply transform() to generate the maximum orders for each user and create a new column "max_orders'
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\adrig\AppData\Local\Temp\ipykernel_13640\3906181915.py:2: FutureWarning: The provided callable <function max at 0x00000295F711F9C0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [58]:
ords_prods_merge['max_order'].value_counts(dropna = False)

max_order
99    1171333
8      811843
6      811396
9      810213
7      803979
5      793140
10     773124
11     769579
4      753543
12     744454
13     736359
14     733970
15     696131
3      686741
17     661981
16     655604
18     621638
19     613711
20     595323
22     593612
21     580515
23     534271
24     529905
25     521960
26     503273
27     487635
28     460949
29     458788
30     458746
31     449436
32     424493
34     418067
33     413854
36     398947
37     374638
40     367311
35     366236
39     357489
41     347926
38     347901
43     337202
44     334794
50     334358
42     331569
47     329459
46     317294
45     316758
49     300430
48     292467
51     285693
53     285458
52     285211
54     240131
56     237279
55     227599
57     200635
58     192981
60     192055
59     188261
61     174191
62     164176
64     161095
63     159384
65     154757
66     133933
67     132211
68     126181
71     122030
73     117901
74     115302
69     111

In [59]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   
15    473747        1             3                   3                 12   
16   2254736        1             4                   4                  7   
17   2254736        1             4                   4                  7   
18   2254736        1             4                   4                  7   
19   2254736        1             4                   4                  7   
20   2254736        1             4                   4                  7   
21    431534        1             5                   4                 15   
22    431534        1             5                   4                 15   
23    431534        1             5                   4                 15   
24    431534        1             5                   4                 15   
25    431534        1             5                   4                 15   
26    431534        1             5                   4                 15   
27    431534        1             5                   4                 15   
28    431534        1             5                   4                 15   
29   3367565        1             6                   2                  7   
30   3367565        1             6                   2                  7   
31   3367565        1             6                   2                  7   
32   3367565        1             6                   2                  7   
33    550135        1             7                   1                  9   
34    550135        1             7                   1                  9   
35    550135        1             7                   1                  9   
36    550135        1             7                   1                  9   
37    550135        1             7                   1                  9   
38   3108588        1             8                   1                 14   
39   3108588        1             8                   1                 14   
40   3108588        1             8                   1                 14   
41   3108588        1             8                   1                 14   
42   3108588        1             8                   1                 14   
43   3108588        1             8                   1                 14   
44   2295261        1             9                   1                 16   
45   2295261        1             9                   1                 16   
46   2295261        1             9                   1                 16   
47   2295261        1             9                   1                 16   
48   2295261        1             9                   1                 16   
49   2295261        1             9                   1                 16   
50   2550362        1     

In [60]:
# Assign the loyalty lebel to the user_id
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [61]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [62]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [64]:
ords_prods_merge.groupby(['loyalty_flag']).user_id.nunique()

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

In [65]:
# Checking new column
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
5         1      New customer             2
6         1      New customer             2
7         1      New customer             2
8         1      New customer             2
9         1      New customer             2
10        1      New customer             2
11        1      New customer             3
12        1      New customer             3
13        1      New customer             3
14        1      New customer             3
15        1      New customer             3
16        1      New customer             4
17        1      New customer             4
18        1      New customer             4
19        1      New customer             4
20        1      New customer             4
21        1      New customer             5
22        1      New customer             5
23        1      New customer             5
24        1      New customer             5
25        1      New customer             5
26        1      New customer             5
27        1      New customer             5
28        1      New customer             5
29        1      New customer             6
30        1      New customer             6
31        1      New customer             6
32        1      New customer             6
33        1      New customer             7
34        1      New customer             7
35        1      New customer             7
36        1      New customer             7
37        1      New customer             7
38        1      New customer             8
39        1      New customer             8
40        1      New customer             8
41        1      New customer             8
42        1      New customer             8
43        1      New customer             8
44        1      New customer             9
45        1      New customer             9
46        1      New customer             9
47        1      New customer             9
48        1      New customer             9
49        1      New customer             9
50        1      New customer            10
51        1      New customer            10
52        1      New customer            10
53        1      New customer            10
54        1      New customer            10
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1

## 03.5 Difference between spending habits of the three types of customers

In [66]:
# mean, min and max of prices grouped by loyalty_flag
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773577  1.0  25.0
New customer      7.801236  1.0  25.0
Regular customer  7.798285  1.0  25.0

The basic statistics of the product prices for each type of customer is very similar. Minimum and maximum prices are the same for the three types of customer. While the lowest mean is for the loyal customer 7.77 , followed by the regular customer and the new customer with the highest mean of 7.8. There´s no a significant difference between the three types of customers. 
I will now look for the information that the price_flag gives.

In [67]:
# look for the frecuency of the price_flag grouped by the loyalty_flag
ords_prods_merge.groupby('loyalty_flag')['price_range_loc'].value_counts(dropna = False)

loyalty_flag      price_range_loc   
Loyal customer    Mid-range product      6950797
                  Low-range product      3214690
                  High-range product      118606
New customer      Mid-range product      4206718
                  Low-range product      1952769
                  High-range product       84503
Regular customer  Mid-range product     10708472
                  Low-range product      4958862
                  High-range product      209442
Name: count, dtype: int64

The consume of products is similar for the three types of customers, being the mid-range products the most consumed among the three types of customer and the high-range products the least consumed. 

## 03.6 Spending flag

In [68]:
# Split data based on user_id, and aggregate the mean of the price
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\adrig\AppData\Local\Temp\ipykernel_13640\3526855396.py:2: FutureWarning: The provided callable <function mean at 0x00000295F712C400> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [69]:
ords_prods_merge['avg_price'].value_counts(dropna = False)

avg_price
8.000000     10340
7.500000      9116
7.750000      7048
7.600000      7042
7.400000      7032
8.200000      7002
8.050000      6718
7.900000      6350
8.300000      6325
7.700000      5417
8.500000      5277
7.100000      5188
7.000000      4943
8.250000      4798
8.099999      4760
7.200000      4726
8.100000      4550
7.875000      4372
7.300000      4306
8.700000      4196
7.700000      4098
8.166667      3930
7.800000      3759
7.800000      3757
7.650000      3734
7.126053      3704
6.683283      3637
7.300000      3584
8.316667      3570
7.450000      3536
8.450000      3530
8.600000      3489
8.066667      3444
7.350000      3286
8.400000      3281
7.250000      3280
8.150000      3178
7.666667      3108
7.441327      3061
7.400000      3061
8.800000      3019
8.125000      3016
6.919312      2936
7.840000      2925
7.735536      2921
7.741157      2906
7.823820      2901
8.233334      2901
7.447783      2888
7.200000      2884
7.633333      2883
7.833333      2883
6.

In [70]:
# Asign spending flag to each user_id
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [71]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [72]:
ords_prods_merge.groupby(['spending_flag']).user_id.nunique()

spending_flag
High spender      3388
Low spender     202821
Name: user_id, dtype: int64

In [73]:
ords_prods_merge[['user_id', 'spending_flag', 'avg_price']].head(60)

user_id spending_flag  avg_price
0         1   Low spender   6.367797
1         1   Low spender   6.367797
2         1   Low spender   6.367797
3         1   Low spender   6.367797
4         1   Low spender   6.367797
5         1   Low spender   6.367797
6         1   Low spender   6.367797
7         1   Low spender   6.367797
8         1   Low spender   6.367797
9         1   Low spender   6.367797
10        1   Low spender   6.367797
11        1   Low spender   6.367797
12        1   Low spender   6.367797
13        1   Low spender   6.367797
14        1   Low spender   6.367797
15        1   Low spender   6.367797
16        1   Low spender   6.367797
17        1   Low spender   6.367797
18        1   Low spender   6.367797
19        1   Low spender   6.367797
20        1   Low spender   6.367797
21        1   Low spender   6.367797
22        1   Low spender   6.367797
23        1   Low spender   6.367797
24        1   Low spender   6.367797
25        1   Low spender   6.367797
26        1   Low spender   6.367797
27        1   Low spender   6.367797
28        1   Low spender   6.367797
29        1   Low spender   6.367797
30        1   Low spender   6.367797
31        1   Low spender   6.367797
32        1   Low spender   6.367797
33        1   Low spender   6.367797
34        1   Low spender   6.367797
35        1   Low spender   6.367797
36        1   Low spender   6.367797
37        1   Low spender   6.367797
38        1   Low spender   6.367797
39        1   Low spender   6.367797
40        1   Low spender   6.367797
41        1   Low spender   6.367797
42        1   Low spender   6.367797
43        1   Low spender   6.367797
44        1   Low spender   6.367797
45        1   Low spender   6.367797
46        1   Low spender   6.367797
47        1   Low spender   6.367797
48        1   Low spender   6.367797
49        1   Low spender   6.367797
50        1   Low spender   6.367797
51        1   Low spender   6.367797
52        1   Low spender   6.367797
53        1   Low spender   6.367797
54        1   Low spender   6.367797
55        1   Low spender   6.367797
56        1   Low spender   6.367797
57        1   Low spender   6.367797
58        1   Low spender   6.367797
59        2   Low spender   7.515897

## 03.7 Frequent vs. non-frequent customers

In [74]:
# Split data based on user_id, and aggregate the median of days_since_last_order
ords_prods_merge['days_last_order_median'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

C:\Users\adrig\AppData\Local\Temp\ipykernel_13640\727747357.py:2: FutureWarning: The provided callable <function median at 0x00000295F7257060> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['days_last_order_median'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)


In [75]:
ords_prods_merge['days_last_order_median'].value_counts(dropna = False)

days_last_order_median
7.0     5682399
6.0     2983604
8.0     2713964
5.0     2486061
4.0     2314289
30.0    1683850
9.0     1635007
3.0     1479987
10.0    1271865
14.0    1213600
13.0    1058299
11.0    1049680
12.0     971198
15.0     744370
2.0      561164
16.0     514568
21.0     421180
17.0     399563
20.0     372214
18.0     356923
19.0     349154
22.0     265844
23.0     201952
28.0     199904
27.0     173220
24.0     169330
26.0     161951
25.0     152639
1.0      152118
29.0     136919
7.5       50062
6.5       45163
8.5       36247
0.0       34240
9.5       33755
10.5      30913
4.5       29496
5.5       28796
12.5      25362
13.5      23823
11.5      19766
14.5      17736
15.5      14732
18.5      14457
17.5      12968
3.5       12468
20.5       9958
16.5       9822
19.5       9416
25.5       7910
22.5       7842
24.5       7412
2.5        7249
21.5       7111
23.5       6656
28.5       6551
27.5       5599
26.5       5514
29.5       5095
1.5        1684
0.5         235
N

In [76]:
# Asign frequency flag to each user_id
ords_prods_merge.loc[ords_prods_merge['days_last_order_median'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [77]:
ords_prods_merge.loc[(ords_prods_merge['days_last_order_median'] > 10) & (ords_prods_merge['days_last_order_median'] <= 20), 'frequency_flag'] = 'Regular customer'

In [78]:
ords_prods_merge.loc[ords_prods_merge['days_last_order_median'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [79]:
ords_prods_merge.groupby(['frequency_flag']).user_id.nunique()

frequency_flag
Frequent customer        86596
Non-frequent customer    59619
Regular customer         59993
Name: user_id, dtype: int64

In [80]:
# Tell Python not to limit the results of head()
pd.options.display.max_rows = None

In [81]:
ords_prods_merge[['user_id', 'frequency_flag', 'days_last_order_median']].head(100)

user_id         frequency_flag  days_last_order_median
0         1  Non-frequent customer                    20.5
1         1  Non-frequent customer                    20.5
2         1  Non-frequent customer                    20.5
3         1  Non-frequent customer                    20.5
4         1  Non-frequent customer                    20.5
5         1  Non-frequent customer                    20.5
6         1  Non-frequent customer                    20.5
7         1  Non-frequent customer                    20.5
8         1  Non-frequent customer                    20.5
9         1  Non-frequent customer                    20.5
10        1  Non-frequent customer                    20.5
11        1  Non-frequent customer                    20.5
12        1  Non-frequent customer                    20.5
13        1  Non-frequent customer                    20.5
14        1  Non-frequent customer                    20.5
15        1  Non-frequent customer                    20.5
16        1  Non-frequent customer                    20.5
17        1  Non-frequent customer                    20.5
18        1  Non-frequent customer                    20.5
19        1  Non-frequent customer                    20.5
20        1  Non-frequent customer                    20.5
21        1  Non-frequent customer                    20.5
22        1  Non-frequent customer                    20.5
23        1  Non-frequent customer                    20.5
24        1  Non-frequent customer                    20.5
25        1  Non-frequent customer                    20.5
26        1  Non-frequent customer                    20.5
27        1  Non-frequent customer                    20.5
28        1  Non-frequent customer                    20.5
29        1  Non-frequent customer                    20.5
30        1  Non-frequent customer                    20.5
31        1  Non-frequent customer                    20.5
32        1  Non-frequent customer                    20.5
33        1  Non-frequent customer                    20.5
34        1  Non-frequent customer                    20.5
35        1  Non-frequent customer                    20.5
36        1  Non-frequent customer                    20.5
37        1  Non-frequent customer                    20.5
38        1  Non-frequent customer                    20.5
39        1  Non-frequent customer                    20.5
40        1  Non-frequent customer                    20.5
41        1  Non-frequent customer                    20.5
42        1  Non-frequent customer                    20.5
43        1  Non-frequent customer                    20.5
44        1  Non-frequent customer                    20.5
45        1  Non-frequent customer                    20.5
46        1  Non-frequent customer                    20.5
47        1  Non-frequent customer                    20.5
48        1  Non-frequent customer                    20.5
49        1  Non-frequent customer                    20.5
50        1  Non-frequent customer                    20.5
51        1  Non-frequent customer                    20.5
52        1  Non-frequent customer                    20.5
53        1  Non-frequent customer                    20.5
54        1  Non-frequent customer                    20.5
55        1  Non-frequent customer                    20.5
56        1  Non-frequent customer                    20.5
57        1  Non-frequent customer                    20.5
58        1  Non-frequent customer                    20.5
59        2       Regular customer                    13.0
60        2       Regular customer                    13.0
61        2       Regular customer                    13.0
62        2       Regular customer                    13.0
63        2       Regular customer                    13.0
64        2       Regular customer                    13.0
65        2       Regular customer                    13.0
66        2       Regular customer                 

In [82]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                    NaN          True       14084                  2   
2                    NaN          True       12427                  3   
3                    NaN          True       26088                  4   
4                    NaN          True       26405                  5   

   reordered  ...    price_range_loc     busiest_day  busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regular days   
1          0  ...  Mid-range product  Regularly busy  Regular days   
2          0  ...  Low-range product  Regularly busy  Regular days   
3          0  ...  Low-range product  Regularly busy  Regular days   
4          0  ...  Low-range product  Regularly busy  Regular days   

   busiest_period_of_day max_order  loyalty_flag avg_price spending_flag  \
0            Most_orders        10  New customer  6.367797   Low spender   
1            Most_orders        10  New customer  6.367797   Low spender   
2            Most_orders        10  New customer  6.367797   Low spender   
3            Most_orders        10  New customer  6.367797   Low spender   
4            Most_orders        10  New customer  6.367797   Low spender   

   days_last_order_median         frequency_flag  
0                    20.5  Non-frequent customer  
1                    20.5  Non-frequent customer  
2                    20.5  Non-frequent customer  
3                    20.5  Non-frequent customer  
4                    20.5  Non-frequent customer  

[5 rows x 24 columns]

In [87]:
ords_prods_merge['user_id'].nunique()

206209

# 04. Export data

In [83]:
# Export ords_prods_merge
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new_var_customer_flags.pkl'))